### ISSUE TITLE
Long latency recorded by canary in the event of interrupted connection to broker

### ISSUE THREAD

**Issue by k-wall** [strimzi/strimzi-canary#133]
If the connection to the broker is interrupted for any reason, internally the canary restores the connection (thanks to sarama retry logic).  This is good, but the latency recorded for the probe message then includes the time taken to reconnect.  This gives a distorted view of message latency behaviour and in our case, leads to spurious alerts.

(In our use-case we use connections.max.reauth.ms, so as sarama doesn't handle [KIP-368](https://cwiki.apache.org/confluence/display/KAFKA/KIP-368) we see frequent disconnections, which exacerbates the issue).

I believe it would be better of this condition were counted as a producer error (or if this behaviour can be made configurable).

**Comment by k-wall** [comment]
There are two cases that cause this issue:

1.  A loss of connection during send.  The sarama retry logic sends the same message after connection is restored, but by then its timestamp is 'stale'.
2. Loss of connection between a successful send and the consume.  In this case the message has made it to the broker, then the connection fails. After reconnection, the message received is already stale.

`config.Producer.Retry.Max = 0` will resolve the first issue, but it won't help the second.  I guess the canary needs some state.

**Comment by tombentley** [comment]
We need to fix Sarama (implement KIP-368) _anyway_, so are these Strimzi workarounds tactical, or would the effort be better spent there?

**Comment by ppatierno** [comment]
I totally agree on implementing KIP-368 on Sarama (I will open an issue there and starting the discussion) but I think that disconnections or producer errors could happen anyway and not related to the re-authentication timeout.
For example, the error related to a broker not being a leader for a specific partition anymore is a transient error that could delay the producer to send the message: producer try to send, get leader not available error, get metadata to update, send to the new broker leader; this will drive to the same problem of higher latency but not related to disconnection and should not be accounted as producer error count. 
Disconnections can even happens during rolling brokers and having higher latency in that case is kind of normal.
I think that latency higher than the usual one should not be a concern until it starts to happens for longer time (in a 5-10 mins window) but if it's just a spike it's part of the Kafka behaviour.

**Comment by ppatierno** [comment]
FYI I opened Sarama https://github.com/Shopify/sarama/issues/2060 to track it and see if community has plan or we should work on our own on this.

**Comment by tombentley** [comment]
> Could we add a digest to the message which is somehow unique to the broker's connection (something based on the ephemeral port number, or a timestamp when the connection was created)? On consumption, the canary should drop messages where the digest doesn't match.

Wouldn't this mean the canary wouldn't notice when it's not able to connect for a protracted period?

I think it's correct for the canary to consider messages "stale" in the cases mentioned, because this is exactly the behaviour any other client would observe. Leaders _do_ change, and that _does_ have an effect on latency, and if that's happening often enough that _is_ a problem, but it's not normally happening enough for it to be problematic. Changes of leadership is the biggest contributor to tail latency, the problem is being able to distinguish a normal latency spike caused by a change of leadership from an abnormal state. Is there any reason why tweaking the timeout for what's considered stale is not the right approach here?

**Comment by k-wall** [comment]
I think the canary should be reporting my 1 and 2 as errors (i.e. increment the producer/consumer error count metric), rather than just recording a long latency that includes  reconnection time.  The user will be able to alert based on those metrics as they see fit.

As @ppatierno mentions there are other cases that cause disconnection apart from KIP-368 so I'd be in favour of a change here even if KIP-368 were supported by the client.

In cases such as leader not available, I think recording a long latency is the right thing.

**Comment by k-wall** [comment]
I took time to reflect on this over the weekend.   I changed my mind on some of the things I said above.

The canary's job is to measure message latency. If there is a problem on the server side that means the a message is slowed (such as brokers being restarted, leader changing) then that should be reflected in the latency figure.

I do think there is good case for turning off producer retry.   At the moment, the latency measured by the canary in the case of a producer error will include the sarama backoff time `Producer.Retry.Backoff` (10ms), which skews the result with a canary concern.  It also hides the fact that there was an unexpected error.

I am suggesting:
`config.Producer.Retry.Max = 0`
as the only change in response to this issue.
